### PR TITLE
FEATURE: Major Overhaul

### PR DESCRIPTION
**Pull request by bwaidelich**
This replaces the usage of the `TYPO3.Party` package in favor of
a new `User` domain model.
This serves as a basis for further adjustments, namely the synchronization
of attributes such that users can update their profiles.

This also comes with some minor tweaks that are not exactly related but
too small to be pushed as separate PRs.

### CODE REVIEW

**Comment by bwaidelich** [comment]
@Sebobo I tried to rewire the party model to achieve our goals, but it was very tedious. So a custom model seems much easier. wdyt?

**Comment by Sebobo** [comment]
@bwaidelich looks good to me.
Would you then want to sync the custom attributes into a serialized field,
or add a field for each attribute?
Would be cool if this would be super dynamic somehow.

**Comment by bwaidelich** [comment]
IMO a property per attribute would work. We'll need to adjust the form to the attributes anyways so making this fully automated would be an overkill IMO.. and with custom properties it's easier to move domain logic into the model (i.e. `getName() { return $this->firstName . ' ' . $this->lastName; }`)
In any case I'd move additional features into a separate PR

**Comment by bwaidelich** [comment]
After talking to @Sebobo again we realized that there is no real need for a persisted user model at all.. We sync it for every authentication anyways and the only place where the `UserRepository` is currently used is within the `indexAction` of the `UserController` to display details of the authenticated user on id.neos.io..

**Comment by Sebobo** [comment]
So should we instead have a short time cache for the data of the logged in user and use that for displaying data which he can update?

**Comment by bwaidelich** [comment]
> So should we instead have a short time cache for the data of the logged in user

Let's see if we run into performance issues, I don't think we need a cache and w/o the sync on every authentication it probably gets even faster

**Comment by bwaidelich** [comment]
This started as a small task to replace the `TYPO3.Party` package, but ended up in a major rewrite that could be squashed to the following commit:

# FEATURE: Major Overhaul

## Main Changes

* Update to Flow 5.1 (was typo3/flow 3.3.19 !)
* Get rid of custom User model and TYPO3.Party
  We fetch the Crowd user anyways for every authentication
* Allow users to update their profile and reset their
  password

## Tweaks

* Rephrase email texts
* Extensive code cleanup (cosmetics, use Value Objects more,
  less accidental complexity, ..)
* Catch exception when trying to authenticate an inactive account
* Verify if email address is already taken (leads to unrecoverable
  errors during discourse sync otherwise)
* Tweak templates
* Cleanup Routes
* Use a proper Authentication Provider to authenticate token requests
  (from Flownative.DoubleOptIn mails)
* Render errors in "CI"

Fixes: #11

File: .gitignore [hunk 1]
```diff
@@ -3,20 +3,17 @@
...
 /Readme.rst
-/Upgrading.txt
 /Upgrading.rst
 /flow
 /flow.bat
-beard.lock
+.editorconfig
 Packages/Application/Neos.IdNeosIo/Resources/Private/Styles/.sass-cache
 Configuration/README
 Configuration/Settings.yaml.example
```
Comment (bwaidelich) [inline .gitignore]: This is the default `.gitignore` added by Flow

File: Packages/Application/Neos.IdNeosIo/Classes/Neos/IdNeosIo/Controller/UserController.php [hunk 2]
```diff
@@ -1,128 +1,156 @@
...
+    public function newAction(): void
+    {
+    }
+
+    /**
+     * @param AddUserDto $addUser
+     * @return void
+     * @throws StopActionException | UnknownPresetException
+     */
+    public function sendActivationEmailAction(AddUserDto $addUser)
```
Comment (kdambekalns) [inline Packages/Application/Neos.IdNeosIo/Classes/Neos/IdNeosIo/Controller/UserController.php]: No void return type here?

File: Packages/Application/Neos.CrowdClient/Migrations/Mysql/Version20160801122505.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
+<?php
+namespace TYPO3\Flow\Persistence\Doctrine\Migrations;
+
+use Doctrine\DBAL\Migrations\AbstractMigration,
+	Doctrine\DBAL\Schema\Schema;
+
+/**
+ * Migrate users from TYPO3.Party (typo3_party_* tables) to new User domain model (neos_crowdclient_domain_model_user)
+ */
+class Version20160801122505 extends AbstractMigration {
```
Comment (bwaidelich) [inline Packages/Application/Neos.CrowdClient/Migrations/Mysql/Version20160801122505.php]: hang on, that's a left-over.. I'll remove it

File: Packages/Application/Neos.CrowdClient/Classes/Neos/CrowdClient/Domain/Dto/User.php [hunk 4]
```diff
@@ -0,0 +1,88 @@
+<?php
+
```
[the remark this answers is not in the document]
Comment (bwaidelich) [reply Packages/Application/Neos.CrowdClient/Classes/Neos/CrowdClient/Domain/Dto/User.php]: Right. PhpStorm prefers it this way.. I'll fix

File: Packages/Application/Neos.IdNeosIo/Classes/Neos/IdNeosIo/Controller/UserController.php (hunk 2 above)
Comment (bwaidelich) [reply Packages/Application/Neos.IdNeosIo/Classes/Neos/IdNeosIo/Controller/UserController.php]: eagle eye karsten

File: Packages/Application/Neos.CrowdClient/Migrations/Mysql/Version20160801122505.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+			$emailAddress = $this->connection->quote($existingUserRow['emailaddress']);
+			$this->addSql("INSERT INTO neos_crowdclient_domain_model_user (persistence_object_identifier, account, firstname, lastname, emailaddress) VALUES ($id, $account, $firstName, $lastName, $emailAddress)");
+		}
+
+		$this->addSql('TRUNCATE TABLE typo3_party_domain_model_person_electronicaddresses_join');
+		$this->addSql('TRUNCATE TABLE typo3_party_domain_model_person');
+		$this->addSql('TRUNCATE TABLE typo3_party_domain_model_personname');
+		$this->addSql('TRUNCATE TABLE typo3_party_domain_model_electronicaddress');
+		$this->addSql('TRUNCATE TABLE typo3_party_domain_model_abstractparty_accounts_join');
+		$this->addSql('TRUNCATE TABLE typo3_party_domain_model_abstractparty');
```
[the remark this answers is not in the document]
Comment (bwaidelich) [reply Packages/Application/Neos.CrowdClient/Migrations/Mysql/Version20160801122505.php]: That migration was a left-over from an earlier version that contained a custom user domain model.
I'll remove the tables manually if everything works ;)

**Comment by kdambekalns** [comment]
🎉

**Pull request merged**